src/gretel_trainer/relational/json.py

```diff
@@ -0,0 +1,310 @@
+    try:
+        obj = load_json(obj)
+    except (ValueError, TypeError, JSONDecodeError):
+        return False
+    else:
+        return isinstance(obj, json_type)
+
+
+def is_dict(obj: Any) -> bool:
+    return is_json(obj, dict)
+
+
+def is_list(obj: Any) -> bool:
+    return isinstance(obj, np.ndarray) or is_json(obj, list)
+
+
+def pandas_json_normalize(series: pd.Series) -> pd.DataFrame:
+    return pd.json_normalize(series.apply(load_json).to_list(), sep=FIELD_SEPARATOR)
+
+
+def nulls_to_empty_dicts(df: pd.DataFrame) -> pd.DataFrame:
+    return df.applymap(lambda x: {} if pd.isnull(x) else x)
+
+
+def nulls_to_empty_lists(series: pd.Series) -> pd.Series:
+    return series.apply(lambda x: x if isinstance(x, list) or not pd.isnull(x) else [])
+
+
+def _normalize_json(
+    nested_dfs: list[tuple[str, pd.DataFrame]], flat_dfs: list[tuple[str, pd.DataFrame]]
+) -> list[tuple[str, pd.DataFrame]]:
+    if not nested_dfs:
+        return flat_dfs
+    name, df = nested_dfs.pop()
+    dict_cols = [
+        col
+        for col in df.columns
+        if df[col].apply(is_dict).any() and df[col].dropna().apply(is_dict).all()
+    ]
+    list_cols = [
+        col
+        for col in df.columns
+        if df[col].apply(is_list).any() and df[col].dropna().apply(is_list).all()
+    ]
+    if dict_cols:
+        df[dict_cols] = nulls_to_empty_dicts(df[dict_cols])
+        for col in dict_cols:
+            new_cols = pandas_json_normalize(df[col]).add_prefix(col + FIELD_SEPARATOR)
+            df = pd.concat([df, new_cols], axis="columns")
+            df = df.drop(columns=new_cols.columns[new_cols.isnull().all()])
+        nested_dfs.append((name, df.drop(columns=dict_cols)))
+    elif list_cols:
+        for col in list_cols:
+            new_table = df[col].explode().dropna().rename(CONTENT_COLUMN).to_frame()
+            new_table[ORDER_COLUMN] = new_table.groupby(level=0).cumcount()
+            nested_dfs.append(
+                (
+                    name + TABLE_SEPARATOR + col,
+                    new_table.reset_index(names=name + ID_SUFFIX),
+                )
+            )
+        nested_dfs.append((name, df.drop(columns=list_cols)))
+    else:
+        flat_dfs.append((name, df))
+    return _normalize_json(nested_dfs, flat_dfs)
+
+
+# Multi-table and multi-column back to single-table with JSON
+
+
+def get_id_columns(df: pd.DataFrame) -> list[str]:
+    return [col for col in df.columns if col.endswith(ID_SUFFIX)]
+
+
+def get_parent_table_name_from_child_id_column(id_column_name: str) -> str:
+    return id_column_name[: -len(ID_SUFFIX)]
+
+
+def get_parent_column_name_from_child_table_name(table_name: str) -> str:
+    return table_name.split(TABLE_SEPARATOR)[-1]
+
+
+def is_child_table(df: pd.DataFrame) -> bool:
+    id_columns = get_id_columns(df)
+    if len(id_columns) != 1:
+        return False
+    id_col = id_columns[0]
+    return all([col in df.columns for col in (id_col, ORDER_COLUMN, CONTENT_COLUMN)])
```

Comment:
Now that you're keeping track of metadata, is there a way to leverage that to make these functions less hacky? E.g. could we use `_is_invented` instead of `is_child_table` without a major refactoring?

Reply:
Good idea, thanks. I was able to remove a few of these entirely, and replace most (but not all) uses of the others. We do need two of these functions _prior to_ creating the `RelationalData` instance that holds all the metadata, but when we're restoring (denormalize) we can leverage the reldata instance.